File: bindings/ruby/examples/style.rb
#!/usr/bin/ruby

require 'storage'

environment = Storage::Environment.new(true)

storage = Storage::Storage.new(environment)
storage.probe()

probed = storage.probed()

# class to define own style
class MyDevicegraphStyleCallbacks < Storage::DevicegraphStyleCallbacks
  def graph()
    attrs = Storage::MapStringString.new()
    attrs['bgcolor'] = 'gray90'
    return attrs
  end

  def nodes
    attrs = Storage::MapStringString.new()
    attrs['shape'] = 'rectangle'
    attrs['style'] = 'filled'
    return attrs
  end

  def node(device)
    attrs = Storage::MapStringString.new()
    attrs['label'] = device.displayname
    attrs['fillcolor'] = Storage.disk?(device) ? 'red' : 'yellow'
    attrs['style'] = 'filled,rounded' if Storage.mount_point?(device)
    return attrs
  end

  def edges()
    attrs = Storage::MapStringString.new()
    return attrs
  end

  def edge(holder)
    attrs = Storage::MapStringString.new()
    attrs['style'] = Storage.subdevice?(holder) ? 'solid' : 'dashed'
    return attrs
  end
end

my_style_callbacks = MyDevicegraphStyleCallbacks.new()

probed.write_graphviz('devicegraph.gv', my_style_callbacks)

system('/usr/bin/dot -Tsvg < devicegraph.gv > devicegraph.svg')
system('/usr/bin/display devicegraph.svg')

File: bindings/python/examples/style.py
#!/usr/bin/python3

# requirements: something to probe


from storage import *
from storageitu import *


set_logger(get_logfile_logger())

environment = Environment(True)

storage = Storage(environment)

storage.probe()

probed = storage.get_probed()


class MyDevicegraphStyleCallbacks(DevicegraphStyleCallbacks):

    def graph(self):
        attrs = MapStringString()
        attrs["bgcolor"] = "gray50:gray70"
        attrs["gradientangle"] = "90"
        return attrs

    def nodes(self):
        attrs = MapStringString()
        attrs["shape"] = "rectangle"
        attrs["style"] = "filled"
        attrs["gradientangle"] = "90"
        return attrs

    def node(self, device):
        attrs = MapStringString()
        attrs["label"] = self.escape(device.get_displayname())
        if is_disk(device):
            attrs["color"] = "coral4"
            attrs["fillcolor"] = "coral3:coral"
            if to_disk(device).is_rotational():
                attrs["image"] = "hdd.svg"
            else:
                attrs["image"] = "ssd.svg"
        elif is_partition(device):
            attrs["color"] = "deepskyblue4"
            attrs["fillcolor"] = "deepskyblue3:deepskyblue"
            attrs["style"] = "filled,dashed"
        elif is_partition_table(device):
            attrs["color"] = "darkolivegreen4"
            attrs["fillcolor"] = "darkolivegreen3:darkolivegreen1"
            attrs["shape"] = "ellipse"
        elif is_filesystem(device):
            attrs["color"] = "aquamarine4"
            attrs["fillcolor"] = "aquamarine3:aquamarine1"
            attrs["style"] = "filled,diagonals"
            if is_nfs(device):
                attrs["image"] = "nfs.svg"
        elif is_mount_point(device):
            attrs["color"] = "gold4"
            attrs["fillcolor"] = "gold3:gold1"
            attrs["style"] = "filled,rounded"
        return attrs

    def edges(self):
        attrs = MapStringString()
        attrs["color"] = "gray20"
        return attrs

    def edge(self, holder):
        attrs = MapStringString()
        if is_subdevice(holder):
            attrs["style"] = "solid"
            attrs["arrowhead"] = "empty"
        else:
            attrs["style"] = "dashed"
            attrs["arrowhead"] = "vee"
        return attrs


my_style_callbacks = MyDevicegraphStyleCallbacks()

probed.write_graphviz("devicegraph.gv", my_style_callbacks)

system("/usr/bin/dot -Tsvg < devicegraph.gv > devicegraph.svg")
system("/usr/bin/display devicegraph.svg")
